src/runtime/proc.go

```diff
@@ -353,6 +353,189 @@ func Gosched() {
+	// is work someone needs to pick up and it might as well be our P. We could,
+	// potentially, directly claim it here and goyield or equivalently to try to
+	// remain on this P, but just parking and letting this P go to findRunnable
+	// avoid duplication of its logic and seems good enough.
+	if !sched.runq.empty() {
+		yieldPark()
+		return
+	}
+
+	// We didn't find anything via cheap O(1) checks of our runq or global runq but
+	// it is possible there are goroutines waiting in runqs of other Ps that are
+	// not being stolen by an idle P -- the lack of idle Ps (npidle=0) is what got
+	// us here. Furthermore, given the lack of idle Ps, it is also possible that
+	// ready conns are waiting for a netpoll to notice them and ready their
+	// goroutines i.e. work to which we should then yield. However, searching all
+	// runqs, and even more so netpoll, is too expensive for every maybeYield
+	// call: being extremely low overhead is essential to allowing Yield() to be
+	// called at high enough frequency to make the caller respond to changing load
+	// promptly.
+	//
+	// Given our main goal here is to reduce/bound *how long* work waits, we can
+	// do more extensive/expensive checks searching all runqs / netpoll less often
+	// so long as we do them often enough. While we can't
+	// define "enough" in term of a number of calls or probabilistic fraction of
+	// calls (e.g. cheaprand()&1023==0) due to variability in caller frequency, we
+	// can frame it in terms of elapsed time: so long as we check for waiting work
+	// after some amount of time has elapsed, we bound how long it waits. We
+	// choose approximately a quarter millisecond for this time: this is long
+	// enough that it should make call overhead negligible, while still being a
+	// duration smaller than the latency of any typical network requests.
+	//
+	// Checking nanotime() every call to implement this cap would in and of itself
+	// be too expensive however, so we instead check the time with an exponential
+	// backoff, using a simple call counter. We combine this counter and the last-
+	// check time in uint32 field on G: 11 lower bits store the counter while the
+	// 21 higher bits store the time as nanos quantized to a 0.25ms "epoch" by
+	// discarding the lower 18 bits of a int64 nanotime() value. When the counter
+	// is 2^k - 1, we check the time; if the 'epoch' has changed, we do the extended
+	// search for waiting work. If the counter is about to overflow but the epoch
+	// hasn't changed, we reset it to half-max to keep checking at the backed-off
+	// rate. Note that while we discard 18 bits to quantize, since the
+	// counter is in the low 11, we only shift by the difference and just mask the
+	// rest out.
+	const yieldCountBits, yieldCountMask = 11, (1 << 11) - 1
+	const yieldEpochShift = 18 - yieldCountBits
+	gp.yieldchecks++
+	// Exp-backoff using 2^k-1 as when we check.
+	if count := gp.yieldchecks & yieldCountMask; (count & (count + 1)) == 0 {
```

Comment:
TIL: about `(count & (count + 1)) == 0. I'm so much more used to `(count & (count - 1)) == 0` which has slightly different but similar behavior.